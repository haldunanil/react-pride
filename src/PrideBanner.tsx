import React from "react";
import styled, { css } from "styled-components";

export interface BannerColors {
  red: string;
  orange: string;
  yellow: string;
  green: string;
  blue: string;
  violet: string;
}

export interface StyledPrideBannerProps
  extends React.HTMLAttributes<HTMLDivElement> {
  /** colors to use for the banner */
  bannerColors?: BannerColors;
  /** determines whether the component should grow in height when the user hovers over it */
  expandOnHover?: boolean;
  /** position for the banner on the page */
  position?: "absolute" | "fixed";
}

export interface PrideBannerProps
  extends Omit<StyledPrideBannerProps, "expandOnHover"> {}

const getHoverCss = ({ expandOnHover }: StyledPrideBannerProps) => {
  if (!expandOnHover) {
    return null;
  }

  return css`
    &:hover {
      height: 50px;
      background-size: 1000% 1000%;
      animation: shift 15s infinite;
    }

    &:hover > span {
      opacity: 1;
    }
  `;
};

export const StyledPrideBanner = styled.div<StyledPrideBannerProps>`
  position: ${(props) => props.position};
  top: 0;
  left: 0;
  right: 0;
  height: 5px;
  background: linear-gradient(
    to right,
    ${(props) => props.bannerColors?.red || "red"},
    ${(props) => props.bannerColors?.orange || "orange"},
    ${(props) => props.bannerColors?.yellow || "yellow"},
    ${(props) => props.bannerColors?.green || "green"},
    ${(props) => props.bannerColors?.blue || "blue"},
    ${(props) => props.bannerColors?.violet || "violet"}
  );
  z-index: 2000;
  transition: height 150ms ease-in-out;

  & > span {
    opacity: 0;
    transition: opacity 150ms ease-in-out;
    display: flex;
    justify-content: center;
    align-items: center;
    height: 100%;
    color: white;
    text-shadow: 0 2px 2px rgba(0, 0, 0, 0.2), 0 0 16px rgba(0, 0, 0, 0.1);
  }

  ${getHoverCss};

  @keyframes shift {
    0% {
      background-position: 0% 0%;
    }

    50% {
      background-position: 100% 100%;
    }

    100% {
      background-position: 0% 0%;
    }
  }
`;

const PrideBanner: React.FC<PrideBannerProps> = ({ children, ...props }) => {
  const expandOnHover = React.Children.count(children) > 0;

  return (
    <StyledPrideBanner {...props} expandOnHover={expandOnHover}>
      <span>{children}</span>
    </StyledPrideBanner>
  );
};

PrideBanner.defaultProps = {
  position: "fixed",
};

export default PrideBanner;
