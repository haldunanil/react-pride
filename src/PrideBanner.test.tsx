import React from "react";
import { shallow } from "enzyme";

import PrideBanner, {
  BannerColors,
  PrideBannerProps,
  StyledPrideBanner,
} from "./PrideBanner";

const positions: PrideBannerProps["position"][] = ["absolute", "fixed"];
const bannerColors: BannerColors = {
  red: "#FF0000",
  orange: "#FFA500",
  yellow: "#FFFF00",
  green: "#008000",
  blue: "#0000FF",
  violet: "#EE82EE",
};

describe(`PrideBanner`, () => {
  let props: PrideBannerProps;

  beforeEach(() => {
    props = {
      children: "We support our LGBTQ+ users!",
    };
  });

  it(`should match snapshot`, () => {
    expect(shallow(<PrideBanner {...props} />)).toMatchSnapshot();
  });

  describe(`StyledPrideBanner`, () => {
    it.each(positions)(
      `should match snapshot when position is '%s'`,
      (position) => {
        expect(
          shallow(<StyledPrideBanner {...props} position={position} />)
        ).toMatchSnapshot();
      }
    );

    it.each([true, false])(
      `should match snapshot when 'expandOnHover' is '%s'`,
      (expandOnHover) => {
        expect(
          shallow(
            <StyledPrideBanner {...props} expandOnHover={expandOnHover} />
          )
        ).toMatchSnapshot();
      }
    );

    it(`should match snapshot when given a 'bannerColors' prop`, () => {
      expect(
        shallow(<StyledPrideBanner {...props} bannerColors={bannerColors} />)
      ).toMatchSnapshot();
    });
  });
});
