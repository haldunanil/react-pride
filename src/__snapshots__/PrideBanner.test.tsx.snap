// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrideBanner StyledPrideBanner should match snapshot when 'expandOnHover' is 'false' 1`] = `
.c0 {
  top: 0;
  left: 0;
  right: 0;
  height: 5px;
  background: linear-gradient( to right, red, orange, yellow, green, blue, violet );
  z-index: 2000;
  -webkit-transition: height 150ms ease-in-out;
  transition: height 150ms ease-in-out;
}

.c0 > span {
  opacity: 0;
  -webkit-transition: opacity 150ms ease-in-out;
  transition: opacity 150ms ease-in-out;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  color: white;
  text-shadow: 0 2px 2px rgba(0,0,0,0.2),0 0 16px rgba(0,0,0,0.1);
}

<div
  className="c0"
>
  We support our LGBTQ+ users!
</div>
`;

exports[`PrideBanner StyledPrideBanner should match snapshot when 'expandOnHover' is 'true' 1`] = `
.c0 {
  top: 0;
  left: 0;
  right: 0;
  height: 5px;
  background: linear-gradient( to right, red, orange, yellow, green, blue, violet );
  z-index: 2000;
  -webkit-transition: height 150ms ease-in-out;
  transition: height 150ms ease-in-out;
}

.c0 > span {
  opacity: 0;
  -webkit-transition: opacity 150ms ease-in-out;
  transition: opacity 150ms ease-in-out;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  color: white;
  text-shadow: 0 2px 2px rgba(0,0,0,0.2),0 0 16px rgba(0,0,0,0.1);
}

.c0:hover {
  height: 50px;
  background-size: 1000% 1000%;
  -webkit-animation: shift 15s infinite;
  animation: shift 15s infinite;
}

.c0:hover > span {
  opacity: 1;
}

<div
  className="c0"
>
  We support our LGBTQ+ users!
</div>
`;

exports[`PrideBanner StyledPrideBanner should match snapshot when given a 'bannerColors' prop 1`] = `
.c0 {
  top: 0;
  left: 0;
  right: 0;
  height: 5px;
  background: linear-gradient( to right, #FF0000, #FFA500, #FFFF00, #008000, #0000FF, #EE82EE );
  z-index: 2000;
  -webkit-transition: height 150ms ease-in-out;
  transition: height 150ms ease-in-out;
}

.c0 > span {
  opacity: 0;
  -webkit-transition: opacity 150ms ease-in-out;
  transition: opacity 150ms ease-in-out;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  color: white;
  text-shadow: 0 2px 2px rgba(0,0,0,0.2),0 0 16px rgba(0,0,0,0.1);
}

<div
  className="c0"
>
  We support our LGBTQ+ users!
</div>
`;

exports[`PrideBanner StyledPrideBanner should match snapshot when position is 'absolute' 1`] = `
.c0 {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  height: 5px;
  background: linear-gradient( to right, red, orange, yellow, green, blue, violet );
  z-index: 2000;
  -webkit-transition: height 150ms ease-in-out;
  transition: height 150ms ease-in-out;
}

.c0 > span {
  opacity: 0;
  -webkit-transition: opacity 150ms ease-in-out;
  transition: opacity 150ms ease-in-out;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  color: white;
  text-shadow: 0 2px 2px rgba(0,0,0,0.2),0 0 16px rgba(0,0,0,0.1);
}

<div
  className="c0"
>
  We support our LGBTQ+ users!
</div>
`;

exports[`PrideBanner StyledPrideBanner should match snapshot when position is 'fixed' 1`] = `
.c0 {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  height: 5px;
  background: linear-gradient( to right, red, orange, yellow, green, blue, violet );
  z-index: 2000;
  -webkit-transition: height 150ms ease-in-out;
  transition: height 150ms ease-in-out;
}

.c0 > span {
  opacity: 0;
  -webkit-transition: opacity 150ms ease-in-out;
  transition: opacity 150ms ease-in-out;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  color: white;
  text-shadow: 0 2px 2px rgba(0,0,0,0.2),0 0 16px rgba(0,0,0,0.1);
}

<div
  className="c0"
>
  We support our LGBTQ+ users!
</div>
`;

exports[`PrideBanner should match snapshot 1`] = `
<styled.div
  expandOnHover={true}
  position="fixed"
>
  <span>
    We support our LGBTQ+ users!
  </span>
</styled.div>
`;
